src/java/org/apache/cassandra/utils/memory/BufferPool.java

```diff
@@ -593,10 +629,11 @@ private void forEach(Consumer<Chunk> consumer)
 
         private void clearForEach(Consumer<Chunk> consumer)
         {
+            int oldCount = count;
             Chunk chunk0 = this.chunk0, chunk1 = this.chunk1, chunk2 = this.chunk2;
-            this.chunk0 = this.chunk1 = this.chunk2 = null;
-            forEach(consumer, count, chunk0, chunk1, chunk2);
             count = 0;
+            this.chunk0 = this.chunk1 = this.chunk2 = null;
+            forEach(consumer, oldCount, chunk0, chunk1, chunk2);
```

Comment:
As written in the commit message, this caused NPE at some point. 
> MicroQueueOfChunks is local to a thread right?

Yes, but this one particular piece has nothing to do with threading. It was simply incorrect even in a single-threaded case, because the callback code could be called during the time the object was in an invalid state (all chunks set to null, but count > 0). And because the callback could call this object again... this led to NPE.

Reply:
Sorry, I missed the commit messages on Friday, and only relied on the PR description. This change looked unrelated, and so I asked about it. I did not disagree with the fix, I just wanted to understand the context for it.